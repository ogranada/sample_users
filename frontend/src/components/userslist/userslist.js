import React, { Component } from "react";
import PropTypes from 'prop-types';
import Swal from 'sweetalert2';
import 'whatwg-fetch';
import { User } from "components/user/user";
import "./userslist.scss";

export class Userslist extends Component {
  constructor(props) {
    super(props);
    this.dataProvider = props.dataProvider;
    this.state = {
      users: []
    };
    this.deleteUser = this.deleteUser.bind(this);
  }

  componentDidMount() {
    return this.updateInformation();
  }

  updateInformation() {
    return this.dataProvider.getUsers()
      .then(users => {
        this.setState({
          users
        });
      });
  }

  deleteUser(user) {
    return () => {
      Swal.fire({
        title: 'Are you sure?',
        text: "You won't be able to revert this!",
        type: 'warning',
        showCancelButton: true,
        confirmButtonColor: '#3085d6',
        cancelButtonColor: '#d33',
        confirmButtonText: 'Yes, delete it!'
      }).then((result) => {
        if (result.value) {
          this.dataProvider.deleteUser(user)
            .then(json => {
              this.updateInformation();
            })
        }
      })
    }
  }

  render() {
    const users = this.state.users;
    return (
      <div className="Userslist">
        <table className="Userslist-table">
          <thead className="Userslist-head">
            <tr>
              <th>Name</th>
              <th className="Userslist-centeredColumn">Status</th>
              <th className="Userslist-centeredColumn">Actions</th>
            </tr>
          </thead>
          <tbody className="Userslist-body">
            {users.map((user, idx) => (
              <tr key={idx}>
                <td>
                  <User user={user} />
                </td>
                <td className="Userslist-centeredColumn">
                  <i className="fa fa-check"></i>
                </td>
                <td className="Userslist-centeredColumn">
                  <button className="Userslist-button">
                    <i className="fa fa-pencil"></i>
                  </button>
                  <button className="Userslist-button Userslist-deleteButton" onClick={this.deleteUser(user)}>
                    <i className="fa fa-trash-o"></i>
                  </button>
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      </div>
    );
  }
}

Userslist.propTypes = {
  dataProvider: PropTypes.any,
};
