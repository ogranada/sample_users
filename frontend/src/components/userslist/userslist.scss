.Userslist {
    padding: 0 0.5em;
    &-table {
        width: 100%;
    }
    &-head {
        text-align: left;
        color: rgb(36, 36, 36);
        line-height: 28px;
    }
    &-body {
        text-align: center;
    }
    &-centeredColumn {
        text-align: center;
    }
    &-button {
        background-color: #eef0f6;
        padding: 0.5em;
        border: none;
        border-radius: 3px;
        margin: 0 0.2em;
        transition: all 0.2s ease;
        &:active {
            background-color: white;
        }
    }

    @media screen and (min-width: 768px) {
        padding-left: 10%;
        padding-right: 10%;
    }

    @media screen and (min-width: 1024px) {
        padding-left: 20%;
        padding-right: 20%;
    }
}