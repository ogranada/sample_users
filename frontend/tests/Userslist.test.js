import React from "react";
import { shallow } from "enzyme";
import { Userslist } from "components/userslist/userslist";

describe("User", () => {
  it("should render correctly", done => {
    const user = {
      firstname: "Oscar",
      lastname: "Granada",
      fullname: "Oscar Granada",
      picture: "http://fake.path"
    };
    const mockDataProvider = {
      getUsers() {
        return Promise.resolve([user, user, user]);
      }
    };
    const component = shallow(<Userslist dataProvider={mockDataProvider} />);
    const instance = component.instance();
    instance.componentDidMount().then(() => {
      expect(component).toMatchSnapshot();
      done();
    });
  });

  it("should delete an item correctly", done => {
    const user = {
      firstname: "Oscar",
      lastname: "Granada",
      fullname: "Oscar Granada",
      picture: "http://fake.path"
    };
    const mockDataProvider = {
      getUsers() {
        return Promise.resolve([user]);
      },
      deleteUser() {
        return Promise.resolve({});
      }
    };
    const component = shallow(<Userslist dataProvider={mockDataProvider} />);
    const instance = component.instance();
    const instance_deleteUser = jest.spyOn(instance, "deleteUser");
    instance.componentDidMount().then(() => {
      expect(component.text()).toContain("<User />");
      component.find(".Userslist-deleteButton").simulate("click");
      expect(instance_deleteUser).toHaveBeenCalled();
      done();
    });
  });
});
