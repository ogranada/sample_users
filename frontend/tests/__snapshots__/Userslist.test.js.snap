// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User should render correctly 1`] = `
<div
  className="Userslist"
>
  <table
    className="Userslist-table"
  >
    <thead
      className="Userslist-head"
    >
      <tr>
        <th>
          Name
        </th>
        <th
          className="Userslist-centeredColumn"
        >
          Status
        </th>
        <th
          className="Userslist-centeredColumn"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody
      className="Userslist-body"
    >
      <tr
        key="0"
      >
        <td>
          <User
            user={
              Object {
                "firstname": "Oscar",
                "fullname": "Oscar Granada",
                "lastname": "Granada",
                "picture": "http://fake.path",
              }
            }
          />
        </td>
        <td
          className="Userslist-centeredColumn"
        >
          <i
            className="fa fa-check"
          />
        </td>
        <td
          className="Userslist-centeredColumn"
        >
          <button
            className="Userslist-button"
          >
            <i
              className="fa fa-pencil"
            />
          </button>
          <button
            className="Userslist-button Userslist-deleteButton"
            onClick={[Function]}
          >
            <i
              className="fa fa-trash-o"
            />
          </button>
        </td>
      </tr>
      <tr
        key="1"
      >
        <td>
          <User
            user={
              Object {
                "firstname": "Oscar",
                "fullname": "Oscar Granada",
                "lastname": "Granada",
                "picture": "http://fake.path",
              }
            }
          />
        </td>
        <td
          className="Userslist-centeredColumn"
        >
          <i
            className="fa fa-check"
          />
        </td>
        <td
          className="Userslist-centeredColumn"
        >
          <button
            className="Userslist-button"
          >
            <i
              className="fa fa-pencil"
            />
          </button>
          <button
            className="Userslist-button Userslist-deleteButton"
            onClick={[Function]}
          >
            <i
              className="fa fa-trash-o"
            />
          </button>
        </td>
      </tr>
      <tr
        key="2"
      >
        <td>
          <User
            user={
              Object {
                "firstname": "Oscar",
                "fullname": "Oscar Granada",
                "lastname": "Granada",
                "picture": "http://fake.path",
              }
            }
          />
        </td>
        <td
          className="Userslist-centeredColumn"
        >
          <i
            className="fa fa-check"
          />
        </td>
        <td
          className="Userslist-centeredColumn"
        >
          <button
            className="Userslist-button"
          >
            <i
              className="fa fa-pencil"
            />
          </button>
          <button
            className="Userslist-button Userslist-deleteButton"
            onClick={[Function]}
          >
            <i
              className="fa fa-trash-o"
            />
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
